torch/onnx/_internal/exporter/_dynamic_shapes.py

```diff
@@ -0,0 +1,372 @@
+    if len(inputs) != len(flat_dynamic_shapes):
+        warnings.warn(
+            "# ONNX model has different number of inputs than the flatten dynamic_shapes. "
+            "The dynamic axes will not be renamed.",
+            UserWarning,
+            stacklevel=3,
+        )
+        return {}
+    rename_mapping: dict[str, str] = {}
+    # NOTE: We assume that the flat_dynamic_shapes is in the same order as the inputs
+    # When the axis is static, or it connects to _DimHint in dynamic shapes, we skip renaming
+    for idx, axes in enumerate(flat_dynamic_shapes):
+        input = inputs[idx]
+        if isinstance(axes, dict):
+            for dim, axis in axes.items():
+                if not isinstance(input.shape[dim], ir.SymbolicDim):
+                    continue
+                old_name = input.shape[dim].value
+                if old_name is None:
+                    continue
+                if isinstance(axis, _DimHint):
+                    continue
+                # NOTE: ExportedProgram could give the axes the same name if they share
+                # the same shape constraints.
+                custom_name = _get_custom_axis_name(axis)
+                if input.shape[dim].value in rename_mapping:
+                    warnings.warn(
+                        f"# The axis name: {custom_name} will not be used, since it shares "
+                        f"the same shape constraints with another axis: {rename_mapping[input.shape[dim].value]}."
+                    )
+                    continue
+                rename_mapping[input.shape[dim].value] = custom_name
+        elif isinstance(axes, (list, tuple)):
+            for dim, axis in enumerate(axes):
+                if not isinstance(input.shape[dim], ir.SymbolicDim):
+                    continue
+                old_name = input.shape[dim].value
+                if old_name is None:
+                    continue
+                if isinstance(axis, _DimHint):
+                    continue
+                # NOTE: ExportedProgram could give the axes the same name if they share
+                # the same shape constraints.
+                custom_name = _get_custom_axis_name(axis)
+                if input.shape[dim].value in rename_mapping:
+                    warnings.warn(
+                        f"# The axis name: {custom_name} will not be used, since it shares "
+                        f"the same shape constraints with another axis: {rename_mapping[input.shape[dim].value]}.",
+                        UserWarning,
+                        stacklevel=3,
+                    )
+                    continue
+                rename_mapping[input.shape[dim].value] = _get_custom_axis_name(axis)
+    return rename_mapping
+
+
+def _get_custom_axis_name(axis: _Dim | str) -> str:
+    """Get the custom axis name from a torch.export.Dim."""
+    if isinstance(axis, _Dim):
+        return axis.__name__
+    return axis
+
+
+def iterate_and_change_axis_names(
```

Comment:
Possible to have unit tests for this function? Also this function is fitting in ir_passes.py

Reply:
Do you recommend to delete the method from ONNXProgram? We can have it as a pass, but we will need to accpet the whole _dynamic_shapes related processings happening inside _compat.py, which is already crowded with step by step validation.

Reply:
I think we can keep the method. I would just mode the model transform part as a pass

Reply:
Done. Now it calls the pass inside the method. I find making a test for this pass requires a lot effort because I need to build a model with all sort of testing conditions. Especially when we already have e2e tests basically guarding the whole functionality. Do you still think we should have a particular test on this? If so, is there an easy way to do it?